iface.go

```diff
@@ -1,23 +1,35 @@
 package nmt
 
-type NamespaceID []byte
-
-type NamespacedData interface {
-	NamespaceID() NamespaceID
-	Data() []byte
-}
-
 type Nmt interface {
+	// NamespaceSize returns the underlying namespace size. Note that
+	// all namespaced data is expected to have the same namespace size.
+	NamespaceSize() int
 	// Push adds data with the corresponding namespace ID to the tree.
-	Push(data NamespacedData)
-	// PushBatch adds a batch of namespaced data.
-	//
-	// Corresponds to several calls of Push() but can be
-	// optimized for one-shot trees  that just take all the input
-	// data and compute the root once.
-	PushBatch(data ...NamespacedData)
+	// Should return an error if the namespace ID size of the input
+	// does not match the tree's NamespaceSize().
+	Push(data NamespacePrefixedData) error
 
 	// Return the namespaced Merkle Tree's root together with the
 	// min. and max. namespace ID.
 	Root() (minNs, maxNS NamespaceID, root []byte)
 }
+
+type NamespacedProver interface {
```

Comment:
Why is this in a separate interface and not part of the tree interface?

Reply:
Anything wrong with that? I prefer to keep interfaces as small as possible and it seems to be good practice: Golang is a bit extreme as many interfaces in the standard library are one-method ifaces only (Reader, Writer, Formatter, CloseNotifier and many more). As a type can implement several interfaces there is no reason to describe all behaviour in one interface. 

I could imagine a optimized version of the tree that does take in a bunch of data and returns the root as quickly as possible. Or, a simplified version of the NMT that is just used to generate diagrams. 
Or, a prover that fetches its data differently than the standard implementation and independently from a NMT instance (e.g. from the network). 

If you feel strongly about this, I can change it.

Reply:
Ah so is NamespacedMerkleTree also a NamespacedProver? So as long as developers don't have to create a separate instance of NamespacedMerkleTree and NamespacedProver to make proofs, that seems OK.

Reply:
No they do not need to create separate instances. It's just wherever in the LL-core code we would expect a NMT we can pass in anything that implements the NMT interface, and if we need a prover, anything that implements the Prover interface. This is only half true though: a prover and verifier actually belongs to a particular tree implementation. There is no real way to communicate this via the go type system (e.g. there is no equivalent to [associated types](https://www.microsoft.com/en-us/research/wp-content/uploads/2005/01/assoc.pdf)).

Note that it is not clear if we even need these interfaces at all! It was just a good way to quickly communicate the public API without having to look scroll through implementation details. I guess keeping them only make sense if we want this library to have contain several different implementations (with the same interface). It might actually be a good idea to remove the interfaces before merging and re-introduce them when we actually see a real need for them...

Reply:
removed the interfaces for now